src/key_value/key_value_server.cc

```diff
@@ -6,21 +6,37 @@
 #include <glog/logging.h>
 
 #include <cstdio>
+#include <fstream>
+#include <iostream>
 #include <string>
 #include <vector>
 
 namespace csci499 {
 
+KeyValueServer::KeyValueServer() : storage_(), storage_file_() {}
+
+KeyValueServer::KeyValueServer(std::string filename)
+    : storage_(), storage_file_(filename) {
+  std::ifstream input(storage_file_);
+  if (input.is_open()) {  // read file if exists and store in snapshot
+    KeyValueSnapshot snapshot;
+    snapshot.ParseFromIstream(&input);
+    storage_.LoadSnapshot(snapshot);
+  }
+  input.close();
+}
+
 Status KeyValueServer::put(ServerContext* context, const PutRequest* request,
                            PutReply* reply) {
   storage_.Put(request->key(), request->value());
   LOG(INFO) << "rpc put key value pair " << request->key() << " "
             << request->value();
+  StoreSnapshot(storage_file_);
```

Comment:
I was actually just talking to Adam about this and you are 100% right that it is an expensive operation.  In a real implementation it would ideally just keep a change long like the example we looked at today, and "compact" the change log periodically.  However, implementing a change log would be fairly hard with removes and everything, so he said it is ok to just call it on every put.  I like your destructor idea as well, but then if a server is unplugged or something disastrous happens there would be no backup.

Reply:
Yeah, I was actually thinking about that during today's lecture when we were talking about snapshots. This seems good to go.